Feature/api 8858 customer sso docs (#964)
livechat/livechat-public-docs: content/authorization/authorizing-api-calls/index.mdx

</Code>
</Section>

# Customer authorization flows

To authorize requests to the [Customer Chat API](/messaging/customer-chat-api/), you'll need a different access token than the one used for the [Agent Chat API](/messaging/agent-chat-api/) and [Configuration API](/management/configuration-api/).
Throughout the whole API documentation, we'll refer to it as `customer_access_token`.

## Cookie grant

> You can receive an identity of a customer. The identity data is stored in cookies: `__lc_cid` and `__lc_cst`. These cookies are protected with `Secure` and `HttpOnly` flags so they are sent only via HTTPS and are accessible only by a specific server. They are valid for 2 years from the first request and are extended in every next call. These cookies are necessary to recognize the customer when an access token is being issued.

Reviewer: ```suggestion
Cookie grant is based on the `__lc_cid` and `__lc_cst` cookies, which store your customer identity data. These cookies are protected with the `Secure` and `HttpOnly` flags, which means that one can send the cookies only via HTTPS and only specific servers can access them.

The cookies are valid for 2 years, and each next call to `https://accounts.livechat.com/customer/token` extends their lifetime. The `__lc_cid` and `__lc_cst` cookies are necessary to recognize a customer when an access token is being issued.
```

</Code>
</Section>

# Customer authorization flows

To authorize requests to the [Customer Chat API](/messaging/customer-chat-api/), you'll need a different access token than the one used for the [Agent Chat API](/messaging/agent-chat-api/) and [Configuration API](/management/configuration-api/).
Throughout the whole API documentation, we'll refer to it as `customer_access_token`.

> Customer authorization flows allow you to receive an identity of a customer. You can do it via [Cookie grant](#cookie-grant) if you want to receive an identity of a single customer or [Agent token grant](#agent-token-grant) in order to manage indentities of multiple customers.

Reviewer: ```suggestion
Customer authorization flows allow you to receive an identity of a customer. You can do it via [Cookie grant](#cookie-grant) if you want to receive an identity of a single customer or via [Agent token grant](#agent-token-grant) in order to manage multiple customer identities.
```

## Cookie grant

Cookie grant is based on the `__lc_cid` and `__lc_cst` cookies, which store your customer identity data. These cookies are protected with the `Secure` and `HttpOnly` flags, which means that one can send the cookies only via HTTPS and only specific servers can access them.

The cookies are valid for 2 years, and each next call to `https://accounts.livechat.com/customer/token` extends their lifetime. The `__lc_cid` and `__lc_cst` cookies are necessary to recognize a customer when an access token is being issued.

> This flow is recommended for front-end apps that support cookies (like an app running in a web browser). Each user of the app receives an identity if a single customer.

Reviewer: ```suggestion
This flow is recommended for frontend apps that support cookies, such as an app running in a web browser. It has its limitations – all app users will share the same customer identity (the one whose data is stored in cookies).
```

Reviewer: I'm not sure if I got the meaning right

Author: Not really - I meant that every client (user) of the frontend app will get its own, unique identity and they will be able to issue an access token using this identity. So there will be many customer identities but each of them will be managed separately by its owner

Reviewer: I understood something different from our conversation earlier 😅 That Cookie grant limits customer identities to 1 because the browser will not store multiple customers' data at the same time.

PS: What do you mean by owner?

Author: In the agent-token flow, owner is the backend integration and it owns many identities. In the cookie flow, owner of the identity is a user of the frontend app, however, there are many users.

Anyway, I have tweaked it a bit. Does it look better, now?

Reviewer: Okay, I think I get it finally :D I'll get rid of that info in brackets because it adds more confusion `(the one whose data is stored in cookies)` than value

Reviewer: I'll also add one short sentence to Agent token auth that this flow is recommended for backend apps, so that it's straightforward

| `expires_in`   | A number in seconds specifying how long the `access_token` will be valid; 28800 sec by default. When it expires, you will need to generate new `access_token` using `refresh_token` (read [Using the refresh token](#using-the-refresh-token) for more details). |
| `token_type`   | Value: `Bearer`                                                                                                                                                                                                                                                  |

> The response also includes important headers with cookies, for example:

Reviewer: ```suggestion
The response also includes headers with the `__lc_cid` and `__lc_cst` cookies, for example:
```

```
Set-Cookie: __lc_cid=3aa138c1-c137-41c6-6b26-cface5857378; Path=/customer; Domain=accounts.livechat.com; Expires=Sun, 18 Jun 2023 12:26:02 GMT; Max-Age=63072000; HttpOnly; Secure; SameSite=None

Set-Cookie: __lc_cst=87b9d0222e63e349da85535d07c81e1e1fb938de41f4e16e1d9ba5965f25db4aeea46b39d0fb; Path=/customer; Domain=accounts.livechat.com; Expires=Sun, 18 Jun 2023 12:26:02 GMT; Max-Age=63072000; HttpOnly; Secure; SameSite=None
```

These cookies will be automatically stored by a web browser and attached to the next requests to acquire new access tokens. 

💡 **Got stuck? See common** [**Errors**](#errors).

</Text>
<Code>

> <CodeSample path={'NEW CUSTOMER'}>

Reviewer: ```suggestion
<CodeSample path={'COOKIE GRANT: NEW CUSTOMER'}>
```

</Code>
</Section>

> ### Step 2: Existing customer

Reviewer: ```suggestion
### Case: Existing customer
```

> If you need a new access token for the identity, you have to make the same request like the one described in Step 1. However, this time, the web browser will attach cookies to the server and you'll be recognized as an existing customer. You'll get a new access token and the cookies will be extended - they will be still valid for 2 years from the last request.

Reviewer: ```suggestion
Use this option to generate a new access token for an already existing customer identity.

To achieve that, make the same exact same request as in [Case: New customer](#case-new-customer-1). 

Your `entity_id`, or any other customer identity data, won' be overwritten. The web browser will attach cookies to the server allowing it to recognize you as an existing customer. You'll get a new access token, and cookies' lifetime will be extended, too. They will be valid for 2 years from your last request to `https://accounts.livechat.com/customer/token`.
```

## Agent token grant

> You can create an identity of a new customer and later use it to perform actions on behalf of the customer. To start with this flow, you'll need an **agent access token** with a scope: `customers:own`. You can get it from one of the [Agent authorization flows](/authorization/authorizing-api-calls/#agent-authorization-flows).

Reviewer: ```suggestion
Agent token grant lets you create a customer identity and use it to act on behalf of a customer. To begin, you'll need an **agent access token** with the `customers:own` scope. You can get it via one of the [Agent authorization flows](/authorization/authorizing-api-calls/#agent-authorization-flows).
```

You may want to use this flow if there's a need to reflect customers and their messages from other systems (like Facebook Messenger or Whatsapp) in LiveChat. An integration that follows this authorization flow behaves like a middleman between these systems and LiveChat.

<Section>
<Text>

> ### Step 1: New customer

Reviewer: ```suggestion
### Case: New customer
```

> First of all, make a HTTP POST request to create a new customer identity by using the following URL:

Reviewer: ```suggestion
First of all, make a HTTP POST request to create a new customer identity by calling the following URL:
```

```url
https://accounts.livechat.com/customer/token
```

> Remember to authorize using agent access token by adding `Authorization` header to the request.

Reviewer: ```suggestion
Remember to authorize your request using the `Authorization` header with an **agent access token**.
```

#### Request

| Parameter       | Required | Description                                                                                                                                                                                                                                                                                        |
|-----------------|----------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `grant_type`    | yes      | Use `agent_token` for this authorization flow                                                                                                                                                                                                                                                      |

Reviewer: ```suggestion
| `grant_type`    | yes      | Use `access_token` (agent access token) for this authorization flow.                                                                                                                                                                                                                                                      |
```

| `client_id`     | yes      | Client ID of your application                                                                                                                                                                                                                                                                      |
| `response_type` | yes      | Value: `token`                                                                                                                                                                                                                                                                                     |

> | `entity_id`     | no       | Customer ID. If you don't specify it, a new customer will be created (Step 1). Specify it (described in Step 2) if you want to [acquire a new token for an existing customer](/getting-started/authorization/faq/#the-access-token-of-my-customer-has-expired-do-i-need-to-create-a-new-customer). |

Reviewer: ```suggestion
| `entity_id`     | no       | Customer ID. If you don't specify it, a new customer identity will be created (Case: New customer). Specify it if you want to [acquire a new token for an existing customer identity](/getting-started/authorization/faq/#the-access-token-of-my-customer-has-expired-do-i-need-to-create-a-new-customer). |
```

Author: I also added a link to the case

| -------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `access_token` | A token you can use to call LiveChat APIs on behalf of the user (customer).                                                                                                                                                                                      |
| `client_id`    | **Client Id** of your app                                                                                                                                                                                                                                        |
| `entity_id`    | The ID of the newly created customer                                                                                                                                                                                                                             |

> | `expires_in`   | A number in seconds specifying how long the `access_token` will be valid; 28800 sec by default. When it expires, you will need to generate new `access_token` using `refresh_token` (read [Using the refresh token](#using-the-refresh-token) for more details). |

Reviewer: ```suggestion
| `expires_in`   | A number in seconds specifying how long your customer `access_token` will be valid; 28800 sec by default. When it expires, you will need to generate new `access_token` using `refresh_token` (read [Using the refresh token](#using-the-refresh-token) for more details). |
```

Author: There was a little mistake here, I changed it to:

A number in seconds specifying how long the `access_token` will be valid; 28800 sec by default. When it expires, you will need to generate new `access_token` following [Case: Existing customer](#case-existing-customer-1)).

| `token_type`   | Value: `Bearer`                                                                                                                                                                                                                                                  |

💡 **Got stuck? See common** [**Errors**](#errors).

</Text>
<Code>

> <CodeSample path={'NEW CUSTOMER'}>

Reviewer: ```suggestion
<CodeSample path={'AGENT TOKEN GRANT: NEW CUSTOMER'}>
```

```shell
curl "https://accounts.livechat.com/customer/token" \
  -H "Authorization: Bearer <access_token>" \
  -X POST
  -d '{
    "grant_type": "agent_token",
    "client_id": "9cbf3a968289727cb3cdfe83ab1d9836",
    "response_type": "token"
      }'
```

</CodeSample>
<CodeResponse title={'Sample response'}>

```json
{
  "access_token": "dal:test_SZhSSbShcZxrv580FA",
  "client_id": "9cbf3a968289727cb3cdfe83ab1d9836",
  "entity_id": "bf18d1a8-1afe-4a3e-4cc0-a3148f1143db",
  "expires_in": 28800,
  "token_type": "Bearer"
}
```

</CodeResponse>
</Code>
</Section>

<Section>
<Text>

> ### Step 2: Existing customer

Reviewer: ```suggestion
### Case: Existing customer
```

> After creating a new customer identity and using it to perform actions on behalf of the customer, you may want to acquire a new access token for the existing customer - the one you have just created. Make the same request as earlier but this time add an additional parameter: `entity_id`. It can be found in the response in Step 1.

Reviewer: ```suggestion
Use this case to get a new access token for an already existing customer identity. 

Make the same request as in [Case: New customer](#case-new-customer-2), but this time, specify `entity_id` to avoid creating a new identity. You can find your `entity_id` [in the response](#response-5) of the request to create a new customer identity.
```

Author: I guess the link should be `#case-new-customer-1`

Reviewer: I'm afraid #case-new-customer-1 will link to Cookie grant https://developers.labs.livechat.com/docs/feature/API-8858-customer-sso-docs/authorization/authorizing-api-calls/#step-1-new-customer

</Text>
<Code>

> <CodeSample path={'EXISTING CUSTOMER'}>

Reviewer: ```suggestion
<CodeSample path={'AGENT TOKEN GRANT: EXISTING CUSTOMER'}>
```

</Code>
</Section>

### Case: Existing customer

Use this option to generate a new access token for an already existing customer identity.

To achieve that, make the same exact same request as in [Case: New customer](#case-new-customer-1). 

Your `entity_id`, or any other customer identity data, won' be overwritten. The web browser will attach cookies to the server allowing it to recognize you as an existing customer. You'll get a new access token, and cookies' lifetime will be extended, too. They will be valid for 2 years from your last request to `https://accounts.livechat.com/customer/token`.

## Agent token grant

Agent token grant lets you create a customer identity and use it to act on behalf of a customer. To begin, you'll need an **agent access token** with the `customers:own` scope. You can get it via one of the [Agent authorization flows](/authorization/authorizing-api-calls/#agent-authorization-flows).

> We recommend this flow if you want to to impersonate customers and simulate them sending messages from other systems, like Facebook Messenger or WhatsApp, in LiveChat. An integration that follows this authorization flow behaves like a middleman between these systems and LiveChat and manages identities of multiple customers.

Reviewer: ```suggestion
We recommend this flow if you want to to impersonate customers and simulate them sending messages from external systems, like Facebook Messenger or WhatsApp, in LiveChat. An integration that follows this authorization flow acts as s middleman between an external system and LiveChat, and it has the ability to manage multiple customer identities.
```

  "expires_in": 28800,
  "token_type": "Bearer"
}
```

</CodeResponse>
</Code>
</Section>

<Section>
<Text>

### Case: Existing customer

Use this case to get a new access token for an already existing customer identity.

Make the same request as in [Case: New customer](#case-new-customer-1), but this time, specify `entity_id` to avoid creating a new identity. You can find your `entity_id` [in the response](#response-5) of the request to create a new customer identity.

</Text>
<Code>
<CodeSample path={'AGENT TOKEN GRANT: EXISTING CUSTOMER'}>

```shell
curl "https://accounts.livechat.com/customer/token" \
  -H "Authorization: Bearer <access_token>" \
  -X POST
  -d '{
    "grant_type": "agent_token",
    "client_id": "9cbf3a968289727cb3cdfe83ab1d9836",
    "response_type": "token",
    "entity_id: "bf18d1a8-1afe-4a3e-4cc0-a3148f1143db"
      }'
```

</CodeSample>
<CodeResponse title={'Sample response'}>

```json
{
  "access_token": "dal:test_SZhSSbShcZxrv580FB",
  "client_id": "9cbf3a968289727cb3cdfe83ab1d9836",
  "entity_id": "bf18d1a8-1afe-4a3e-4cc0-a3148f1143db",
  "expires_in": 28800,
  "token_type": "Bearer"
}
```

</CodeResponse>
</Code>
</Section>

## Revoking tokens

> Look at the [Customer Accounts API specification](/authorization/customer-accounts-api/#tag/Tokens/paths/~1token/delete) to find more details about revoking tokens.

Reviewer: ```suggestion
Look at the [Customer Accounts API specification](/authorization/customer-accounts-api/#tag/Tokens/paths/~1token/delete) to find more details about revoking customer tokens.
```

## Validating the access token

> The instructions on how to validate access token of a customer are available in [Customer Accounts API specification](/authorization/customer-accounts-api/#tag/Tokens/paths/~1info/get).

Reviewer: ```suggestion
The instruction on how to validate customer access token of a customer is available in the [Customer Accounts API specification](/authorization/customer-accounts-api/#tag/Tokens/paths/~1info/get).
```
